etp-core/etp-backend/src/main/clj/solita/etp/service/energiatodistus_pdf.clj

```diff
@@ -866,22 +873,86 @@
-                (file-service/upsert-file-from-bytes aws-s3-client
-                                                     key))
-           filename)))))
+
+(defn sign-energiatodistus-pdf
+  ([db aws-s3-client whoami now id language signature-and-chain]
+   (sign-energiatodistus-pdf db aws-s3-client whoami now id language signature-and-chain :mpollux))
+  ([db aws-s3-client whoami now id language
+    {:keys [chain] :as signature-and-chain} signing-method]
+   (when-let [energiatodistus
+              (energiatodistus-service/find-energiatodistus db id)]
+     (do-when-signing
+       energiatodistus
+       #(do
+          (validate-certificate! (:sukunimi whoami)
+                                 now
+                                 (first chain)
+                                 signing-method)
+          (let [key (energiatodistus-service/file-key id language)
+                content (file-service/find-file aws-s3-client key)
+                content-bytes (.readAllBytes content)
+                pkcs7 (puumerkki/make-pkcs7 signature-and-chain content-bytes)
+                filename (str key ".pdf")
+                ]
+            (->> (write-signature! id language content-bytes pkcs7)
+                 (file-service/upsert-file-from-bytes aws-s3-client
+                                                      key))
+            filename))))))
+
+(defn cert-pem->one-liner-without-headers [cert-pem]
+  "Given a certificate in PEM format `cert-pem` removes
+  headers and linebreaks from it."
+  (-> cert-pem
+      (str/replace #"-----BEGIN CERTIFICATE-----" "")
+      (str/replace #"-----END CERTIFICATE-----" "")
+      (str/replace #"\n" "")))
+
+(def cert-chain-three-long-leaf-first
+  (let [leaf config/system-signature-certificate-leaf
+        intermediate config/system-signature-certificate-intermediate
+        root config/system-signature-certificate-root]
+    (mapv cert-pem->one-liner-without-headers [leaf intermediate root])))
+
+(defn- data->signed-digest [data aws-kms-client]
+  (->> data
+       (sign-service/sign aws-kms-client)
+       (.readAllBytes)))
+
+(defn- cancel-sign-with-system
+  [db whoami id]
+  (println "Cancelling the signing!")
+  (energiatodistus-service/cancel-energiatodistus-signing! db whoami id))
+
+(defn sign-with-system
+  "Does the whole process of sig"
+  [{:keys [db whoami now id language aws-s3-client aws-kms-client]}]
+  (println "Starting signing with system")
+  (let [start-response (energiatodistus-service/start-energiatodistus-signing! db whoami id)]
+    (if (not (= :ok start-response))
+      (do (println "Could not start the signing ( " start-response ")")
+          ;; If failed to start the signing it should not be cancelled either.
+          start-response)
+      (let [digest-response (find-energiatodistus-digest db aws-s3-client id language)]
+        (if (or (= :already-signed digest-response) (= :not-in-signing digest-response) (nil? digest-response))
```

Comment:
Ehdon voisi eriyttää omaan funktioonsa ja näin saisi helpommin luettavaksi tämän